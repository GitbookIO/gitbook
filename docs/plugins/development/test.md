# Testing your plugin

### On local machine

Testing your plugin on your book before publishing is possible using [npm link](https://docs.npmjs.com/cli/link).

In the plugin's folder, run:

```
$ npm link
```

then in your book's folder:

```
$ npm link gitbook-plugin-<plugin's name>
```

### Unit testing on Travis

[gitbook-tester](https://github.com/todvora/gitbook-tester) makes it easy to write **Node.js/Mocha** unit tests for your plugins. Using [Travis.org](https://travis.org), tests can be run on each commit/tag.
