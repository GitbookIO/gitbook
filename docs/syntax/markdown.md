# Markdown

Most of the examples from this documentation are in Markdown. Markdown is default parser for GitBook, but one can also opt for the [AsciiDoc syntax](asciidoc.md).

Here’s an overview of Markdown syntax that you can use with GitBook (same as GitHub with some additions).

### Headings

To create a heading, add one to six `#` symbols before your heading text. The number of # you use will determine the size of the heading.

```markdown
# This is an <h1> tag
## This is an <h2> tag
###### This is an <h6> tag
```

GitBook supports a nice way for explicitly setting the header ID. If you follow the header text with an opening curly bracket (separated from the text with a least one space), a hash, the ID and a closing curly bracket, the ID is set on the header. If you use the trailing hash feature of atx style headers, the header ID has to go after the trailing hashes. For example:

```markdown
Hello {#id}
-----

# Hello {#id}

# Hello # {#id}
```

### Paragraphs and Line Breaks {#paragraphs}

A paragraph is simply one or more consecutive lines of text, separated by one or more blank lines. (A blank line is any line that looks like a blank line — a line containing nothing but spaces or tabs is considered blank.) Normal paragraphs should not be indented with spaces or tabs.

```
Here's a line for us to start with.

This line is separated from the one above by two newlines, so it will be a *separate paragraph*.
```

### Emphasis {#emphasis}

```markdown
*This text will be italic*
_This will also be italic_

**This text will be bold**
__This will also be bold__

~~This text will be crossed out.~~

_You **can** combine them_
```

### Lists {#lists}

Markdown supports ordered (numbered) and unordered (bulleted) lists.

##### Unordered

Unordered lists use asterisks, pluses, and hyphens — interchangably — as list markers:

```markdown
* Item 1
* Item 2
  * Item 2a
  * Item 2b
```

##### Ordered

Ordered lists use numbers followed by periods:

```markdown
1. Item 1
2. Item 2
3. Item 3
   * Item 3a
   * Item 3b
```

### Links {#links}

Markdown supports two style of links: inline and reference.

A simple link can be created by surrounding the text with square brackets and the link URL with parentheses:

```markdown
This is [an example](http://example.com/ "Title") inline link with a title.

[This link](http://example.net/) has no title attribute.
```

Links can point to relative paths, anchors or absolute urls.


### References

There is another way to create links which does not interrupt the text flow. The URL and title are defined using a reference name and this reference name is then used in square brackets instead of the link URL:

```markdown
This is [an example][id] reference-style link.
```

Then, anywhere in the document, you define your link label like this, on a line by itself:

```markdown
[id]: http://example.com/  "Optional Title Here"
```

### Images {#images}

Images can be created in a similar way than links: just use an exclamation mark before the square brackets. The link text will become the alternative text of the image and the link URL specifies the image source:

```markdown
An image: ![gras](img/image.jpg)
```

### Blockquotes {#blockquotes}

A blockquote is started using the `>` marker followed by an optional space; all following lines that are also started with the blockquote marker belong to the blockquote. You can use any block-level elements inside a blockquote:

```markdown
As Kanye West said:

> We're living the future so
> the present is our past.
```

### Tables {#tables}

You can create tables by assembling a list of words and dividing them with hyphens `-` (for the first row), and then separating each column with a pipe `|`:

```markdown
| First Header  | Second Header |
| ------------- | ------------- |
| Content Cell  | Content Cell  |
| Content Cell  | Content Cell  |
```

The pipes on either end of the table are optional. Cells can vary in width and do not need to be perfectly aligned within columns. There must be at least three hyphens in each column of the header row.

### Code {#code}

Markdown supports two different code block styles. One uses lines indented with either four spaces or one tab whereas the other uses lines with tilde characters as delimiters – therefore the content does not need to be indented:

```markdown
This is a sample code block.

    Continued here.

```

##### Fenced code blocks

You can create fenced code blocks by placing triple backticks ` ``` ` before and after the code block. We recommend placing a blank line before and after code blocks to make the raw formatting easier to read.

    ```
    function test() {
      console.log("notice the blank line before this function?");
    }
    ```

##### Syntax highlighting

You can add an optional language identifier to enable syntax highlighting in your fenced code block.

For example, to syntax highlight Ruby code:

    ```ruby
    require 'redcarpet'
    markdown = Redcarpet.new("Hello World!")
    puts markdown.to_html
    ```

##### Inline code

Text phrases can be marked up as code by surrounding them with backticks:

    Use `gitbook` to convert the `text` in markdown
    syntax to HTML.

### Footnotes

GitBook supports a simple syntax for such footnotes. Footnotes are relative to each pages.

```markdown
Text prior to footnote reference.[^2]

[^2]: Comment to include in footnote.
```

### HTML

GitBook supports use of raw HTML in your text, Markdown syntax in HTML is not processed:

```
<div>
Markdown here will not be **parsed**
</div>
```

### Horizontal Rule

Horizontal Rules can be inserted using three or more asterisks, dashes or underscores, optionally separated by spaces or tabs, on an otherwise blank line:

```markdown
Three or more...

---

Hyphens

***

Asterisks

```

### Ignoring Markdown formatting

You can tell GitBook to ignore (or escape) Markdown formatting by using `\` before the Markdown character.

```
Let's rename \*our-new-project\* to \*our-old-project\*.
```
