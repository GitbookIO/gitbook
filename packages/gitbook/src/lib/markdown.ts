import rehypeSanitize from 'rehype-sanitize';
import rehypeStringify from 'rehype-stringify';
import remarkGfm from 'remark-gfm';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import { unified } from 'unified';

/**
 * Parse markdown and output HTML.
 */
export async function parseMarkdown(markdown: string): Promise<string> {
    const file = await unified()
        .use(remarkParse)
        .use(remarkGfm)
        .use(remarkRehype)
        .use(rehypeSanitize)
        .use(rehypeStringify)
        .process(markdown);

    return file.toString();
}
