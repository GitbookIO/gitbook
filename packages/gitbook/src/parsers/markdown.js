const { State } = require('markup-it');
const markdown = require('markup-it/lib/markdown');
const html = require('markup-it/lib/html');
const annotateCodeBlocks = require('./annotateCodeBlocks');

const FILE_EXTENSIONS = [
    '.md',
    '.markdown',
    '.mdown'
];

/**
 * Render a document as markdown.
 * @param  {Document} document
 * @return {String} text
 */
function toText(document) {
    const state = State.create(markdown);
    return state.serializeDocument(document);
}

/**
 * Parse markdown into a document.
 * @param  {String} text
 * @return {Document} document
 */
function toDocument(text) {
    const state = State.create(markdown);
    return state.deserializeToDocument(text);
}

/**
 * Prepare a document for parsing
 * @param  {String} text
 * @return {String} text
 */
function prepare(text) {
    let doc = toDocument(text);
    doc = annotateCodeBlocks(doc);
    return toText(doc);
}

/**
 * Render markdown to HTML.
 * @param  {String} text
 * @return {String} html
 */
function toHTML(text) {
    const document = toDocument(text);
    const state = State.create(html);

    return state.serializeDocument(document);
}

/**
 * Render markdown to inline HTML.
 * @param  {String} text
 * @return {String} html
 */
function toInlineHTML(text) {
    const document = toDocument(text);
    const state = State.create(html);

    return state.serializeDocument(document);
}

module.exports = {
    name: 'markdown',
    FILE_EXTENSIONS,
    prepare,
    toText,
    toDocument,
    toHTML,
    toInlineHTML
};
