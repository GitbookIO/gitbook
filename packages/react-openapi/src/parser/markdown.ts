import { AnyObject } from '@scalar/openapi-parser';
import { traverse } from './traverse';

/**
 * Parse descriptions in the spec to markdown.
 */
export async function parseDescriptions<T extends AnyObject>(input: {
    specification: T;
    parseMarkdown: (input: string) => Promise<string>;
}): Promise<T> {
    const { specification, parseMarkdown } = input;
    const promises: Record<string, Promise<string>> = {};
    const results: Record<string, string> = {};
    traverse(specification, (obj, path) => {
        if (checkHasDescription(obj) && path) {
            promises[hashPath(path)] = parseMarkdown(obj.description);
        }
        return obj;
    });
    await Promise.all(
        Object.entries(promises).map(async ([key, promise]) => {
            results[key] = await promise;
        }),
    );
    return traverse(specification, (obj, path) => {
        if (
            checkHasDescription(obj) &&
            typeof obj.description === 'string' &&
            path &&
            results[hashPath(path)]
        ) {
            obj.description = results[hashPath(path)];
        }
        return obj;
    }) as T;
}

/**
 * Check if the object contains a description.
 */
function checkHasDescription(obj: AnyObject): obj is { description: string } {
    return 'description' in obj && typeof obj.description === 'string';
}

/**
 * Hash a path.
 */
function hashPath(path: string[]): string {
    return path.join('/');
}
