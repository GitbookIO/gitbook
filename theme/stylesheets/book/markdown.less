.book.color-theme-1 .book-body .page-wrapper .page-inner section.normal {
    color:@page-color-1;

    pre, code {
        background: #fdf6e3;
        color: #657b83;
        border-color: darken(#fdf6e3, 15%);

        @import "./highlight/sepia.less";
    }
}
.book.color-theme-2 .book-body .page-wrapper .page-inner section.normal {
    color:@page-color-2;

    pre, code {
        background: black;
        color: #eaeaea;
        border-color: #000;

        @import "./highlight/night.less";
    }
}
.book .book-body .page-wrapper .page-inner section.normal {
    padding-bottom: 25px;
    padding-top: 15px;
    color:@page-color;

    & > *:first-child {
        margin-top: 0 !important; }
    & > *:last-child {
        margin-bottom: 0 !important; }

    a {
        color: #4183C4; }
    a.absent {
        color: #cc0000; }
    a.anchor {
        display: block;
        width: 20px;
        cursor: pointer;
        position: absolute;
        top: 0;
        left: -20px;
        bottom: 0;
        outline: none;

        span {
            font-size: 13px;
            line-height: 13px;
            color: #000;
            display: none;
        }

        &:hover {
            span {
                display: inline-block;
            }
        }
    }

    h1, h2, h3, h4, h5, h6 {
        line-height: 1.7;
        margin: 20px 0 10px;
        padding: 0;
        font-weight: bold;
        -webkit-font-smoothing: antialiased;
        cursor: text;
        position: relative;

        &:hover {
            a.anchor span {
                display: inline-block;
            }
        }
    }

    h1 tt, h1 code {
        font-size: inherit; }

    h2 tt, h2 code {
        font-size: inherit; }

    h3 tt, h3 code {
        font-size: inherit; }

    h4 tt, h4 code {
        font-size: inherit; }

    h5 tt, h5 code {
        font-size: inherit; }

    h6 tt, h6 code {
        font-size: inherit; }

    h1 {
        font-size: 2.5em;
    }

    h2 {
        font-size: 2em;
        border-bottom: 1px solid #eee;
    }

    h3 {
        font-size: 1.5em; }

    h4 {
        font-size: 1.2em; }

    h5 {
        font-size: 1em; }

    h6 {
        font-size: 1em; }

    p, blockquote, ul, ol, dl, table, pre {
        margin: 15px 0; }

    hr {

    }

    body > h2:first-child {
        margin-top: 0;
        padding-top: 0; }
    body > h1:first-child {
        margin-top: 0;
        padding-top: 0; }
        body > h1:first-child + h2 {
            margin-top: 0;
            padding-top: 0; }
    body > h3:first-child, body > h4:first-child, body > h5:first-child, body > h6:first-child {
        margin-top: 0;
        padding-top: 0; }

    a:first-child h1, a:first-child h2, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
        margin-top: 0;
        padding-top: 0; }

    h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
        margin-top: 0; }

    li p.first {
        display: inline-block; }

    ul, ol {
        padding-left: 30px; }

    ul :first-child, ol :first-child {
        margin-top: 0; }

    ul :last-child, ol :last-child {
        margin-bottom: 0; }

    ul p {
        margin: 0px;
    }
    ul ul {
        margin: 0px;
    }

    dl {
        padding: 0; }
        dl dt {
            font-size: 14px;
            font-weight: bold;
            font-style: italic;
            padding: 0;
            margin: 15px 0 5px; }
            dl dt:first-child {
                padding: 0; }
            dl dt > :first-child {
                margin-top: 0; }
            dl dt > :last-child {
                margin-bottom: 0; }
        dl dd {
            margin: 0 0 15px;
            padding: 0 15px; }
            dl dd > :first-child {
                margin-top: 0; }
            dl dd > :last-child {
                margin-bottom: 0; }

    blockquote {
        border-left: 4px solid #dddddd;
        padding: 0 15px;
        color: #777777;

        p {
            font-size: inherit;
        }
    }
        blockquote > :first-child {
            margin-top: 0;
        }
        blockquote > :last-child {
            margin-bottom: 0; }

    table {
        width: 100%;
        padding: 0;
    }

        table tr {
            border-top: 1px solid #cccccc;
            background-color: white;
            margin: 0;
            padding: 0; }
            table tr:nth-child(2n) {
                background-color: #f8f8f8; }
            table tr th {
                font-weight: bold;
                border: 1px solid #cccccc;
                text-align: left;
                margin: 0;
                padding: 6px 13px; }
            table tr td {
                border: 1px solid #cccccc;
                text-align: left;
                margin: 0;
                padding: 6px 13px; }
            table tr th :first-child, table tr td :first-child {
                margin-top: 0; }
            table tr th :last-child, table tr td :last-child {
                margin-bottom: 0; }

    img {
        max-width: 100%;
        display: block;
        margin: 0px auto;
    }

    span.frame {
        display: block;
        overflow: hidden; }
        span.frame > span {
            border: 1px solid #dddddd;
            display: block;
            float: left;
            overflow: hidden;
            margin: 13px 0 0;
            padding: 7px;
            width: auto; }
        span.frame span img {
            display: block;
            float: left; }
        span.frame span span {
            clear: both;
            color: #333333;
            display: block;
            padding: 5px 0 0; }
    span.align-center {
        display: block;
        overflow: hidden;
        clear: both; }
        span.align-center > span {
            display: block;
            overflow: hidden;
            margin: 13px auto 0;
            text-align: center; }
        span.align-center span img {
            margin: 0 auto;
            text-align: center; }
    span.align-right {
        display: block;
        overflow: hidden;
        clear: both; }
        span.align-right > span {
            display: block;
            overflow: hidden;
            margin: 13px 0 0;
            text-align: right; }
        span.align-right span img {
            margin: 0;
            text-align: right; }
    span.float-left {
        display: block;
        margin-right: 13px;
        overflow: hidden;
        float: left;

        & span {
            margin: 13px 0 0;
        }
    }

    span.float-right {
        display: block;
        margin-left: 13px;
        overflow: hidden;
        float: right;

        & > span {
            display: block;
            overflow: hidden;
            margin: 13px auto 0;
            text-align: right;
        }
    }

    code, tt {
        margin: 0 2px;
        padding: 0 5px;
        white-space: nowrap;
        border: 1px solid #eaeaea;
        background-color: #f8f8f8;
        border-radius: 3px;
    }

    pre code {
        margin: 0;
        padding: 0;
        white-space: pre;
        border: none;
        background: transparent;

        @import "./highlight/white.less";
    }

    pre {
        font-size: inherit;
        line-height: 1.5em;
        overflow: auto;
        padding: 20px;
        border-radius: 3px;
    }

    pre code, pre tt {
        background-color: transparent;
        border: none;
    }
}
