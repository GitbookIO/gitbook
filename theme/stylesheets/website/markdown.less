.book .book-body .page-wrapper .page-inner section {
    display: none;
}

.book .book-body .page-wrapper .page-inner section.normal {
    .markdown-content(@content-color, @content-line-height);

    pre, code {
        @import "./highlight/white.less";
    }

    .glossary-term {
        cursor: help;
        text-decoration: underline;
    }
}

.book.color-theme-1 .book-body .page-wrapper .page-inner section.normal {
    color:@page-color-1;

    pre, code {
        background: #fdf6e3;
        color: #657b83;
        border-color: darken(#fdf6e3, 15%);

        @import "./highlight/sepia.less";
    }

    table {
        th, td {
            border-color: darken(#fdf6e3, 25%);
        }

        tr {
            background: #fdf6e3;
            color: inherit;
            border-color: #444;
        }

        tr:nth-child(2n) {
            background-color: darken(#fdf6e3, 5%);
        }
    }
}

.book.color-theme-2 .book-body .page-wrapper .page-inner section.normal {
    color:@page-color-2;

    pre, code {
        background: black;
        color: #eaeaea;
        border-color: #000;

        @import "./highlight/night.less";
    }

    table {
        th, td {
            border-color: #444;
        }

        tr {
            background-color: black;
            color: #eaeaea;
            border-color: #444;
        }

        tr:nth-child(2n) {
            background-color: #222;
        }
    }
}
