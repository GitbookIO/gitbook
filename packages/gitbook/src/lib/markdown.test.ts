import { describe, expect, it } from 'bun:test';

import { parseMarkdown } from './markdown';

describe('parseMarkdown', () => {
    it('should parse a simple table', async () => {
        const result = await parseMarkdown(`## Table

| a | b  |  c |  d  |
| - | :- | -: | :-: |`);

        expect(result).toMatchSnapshot();
    });

    it('should parse a complex table', async () => {
        const result =
            await parseMarkdown(`Returns information for all non-fungible tokens for an account.

## Ordering
When considering NFTs, their order is governed by a combination of their numerical **token.Id** and **serialnumber** values, with **token.id** being the parent column.
A serialnumbers value governs its order within the given token.id

In that regard, if a user acquired a set of NFTs in the order (2-2, 2-4 1-5, 1-1, 1-3, 3-3, 3-4), the following layouts illustrate the ordering expectations for ownership listing
1. **All NFTs in ASC order**: 1-1, 1-3, 1-5, 2-2, 2-4, 3-3, 3-4
2. **All NFTs in DESC order**: 3-4, 3-3, 2-4, 2-2, 1-5, 1-3, 1-1
3. **NFTs above 1-1 in ASC order**: 1-3, 1-5, 2-2, 2-4, 3-3, 3-4
4. **NFTs below 3-3 in ASC order**: 1-1, 1-3, 1-5, 2-2, 2-4
5. **NFTs between 1-3 and 3-3 inclusive in DESC order**: 3-4, 3-3, 2-4, 2-2, 1-5, 1-3

Note: The default order for this API is currently DESC

## Filtering
When filtering there are some restrictions enforced to ensure correctness and scalability.

**The table below defines the restrictions and support for the NFT ownership endpoint**

| Query Param   | Comparison Operator | Support | Description           | Example |
| ------------- | ------------------- | ------- | --------------------- | ------- |
| token.id      | eq                  | Y       | Single occurrence only. | ?token.id=X |
|               | ne                  | N       | | |
|               | lt(e)               | Y       | Single occurrence only. | ?token.id=lte:X |
|               | gt(e)               | Y       | Single occurrence only. | ?token.id=gte:X |
| serialnumber  | eq                  | Y       | Single occurrence only. Requires the presence of a **token.id** query | ?serialnumber=Y |
|               | ne                  | N       | | |
|               | lt(e)               | Y       | Single occurrence only. Requires the presence of an **lte** or **eq** **token.id** query | ?token.id=lte:X&serialnumber=lt:Y |
|               | gt(e)               | Y       | Single occurrence only. Requires the presence of an **gte** or **eq** **token.id** query | ?token.id=gte:X&serialnumber=gt:Y |
| spender.id    | eq                  | Y       | | ?spender.id=Z |
|               | ne                  | N       | | |
|               | lt(e)               | Y       | | ?spender.id=lt:Z |
|               | gt(e)               | Y       | | ?spender.id=gt:Z |

Note: When searching across a range for individual NFTs a **serialnumber** with an additional **token.id** query filter must be provided.
Both filters must be a single occurrence of **gt(e)** or **lt(e)** which provide a lower and or upper boundary for search.`);

        expect(result).toMatchSnapshot();
    });

    it('should parse html', async () => {
        const result = await parseMarkdown(
            '<div style="\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  margin-bottom: 1rem;\n">\n  <div style="\n  display: flex;\n  align-items: center;\n  font-family: monospace;\n  font-weight: 800;\n">\n  <div style="color: #EC7063; margin-right: 0.3rem; font-size: 16px;">\n    ID:\n  </div>\n  <div style="color: #515A5A; font-size: 14px;">\n    get_trending_events\n  </div>\n</div>\n\n  <div>\n  <div>\n<span style="\n  border-radius: 10px;\n  padding-top: 0.25rem;\n  padding-bottom: 0.25rem;\n  padding-left: 0.5rem;\n  padding-right: 0.5rem;\n  margin-right: 1rem;\n  background-color: rgb(16, 185, 129);\n  display: flex;\n  align-items: center;\n">\n  <span style="fill: #fff; height: 1em; width: 1em; display: inline-flex; margin-right: 0.4rem;">\n    <svg viewBox="0 0 1792 1792" xmlns="http://www.w3.org/2000/svg">\n      <path d="M1376 768q40 0 68 28t28 68v576q0 40-28 68t-68 28h-960q-40 0-68-28t-28-68v-576q0-40 28-68t68-28h32v-320q0-185 131.5-316.5t316.5-131.5 316.5 131.5 131.5 316.5q0 26-19 45t-45 19h-64q-26 0-45-19t-19-45q0-106-75-181t-181-75-181 75-75 181v320h736z"/>\n</svg>\n  </span>\n  <span style="font-weight: 500; font-family: monospace; color: #fff;">\n    sports_events_trending\n  </span>\n</span>\n</div>\n\n</div>\n\n</div>\n<div style="display: flex; margin-bottom: 0.5rem; font-size: small;">\n  <span style="\n  border-radius: 20px;\n  border-style: solid;\n  padding-top: 0.15rem;\n  padding-bottom: 0.15rem;\n  padding-left: 0.45rem;\n  padding-right: 0.45rem;\n  margin-right: 0.5rem;\n  border-color: #1D8348;\n  color: #1D8348;\n  align-items: center;\n  font-weight: 500;\n">\n    Sports\n</span>\n<span style="\n  border-radius: 20px;\n  border-style: solid;\n  padding-top: 0.15rem;\n  padding-bottom: 0.15rem;\n  padding-left: 0.45rem;\n  padding-right: 0.45rem;\n  margin-right: 0.5rem;\n  border-color: #515A5A;\n  color: #515A5A;\n  align-items: center;\n  font-weight: 500;\n">\n    Personalization\n</span>\n\n</div>\n\nReturns trending events for the given time frame.\n\nThe events are sorted by `trending score`.\n\n<div style="\n  border-left: 5px solid rgb(0,184,212);\n  border-radius: .1rem;\n  box-shadow: 0 0.2rem 0.5rem rgb(0 0 0 / 5%), 0 0.025rem 0.05rem rgb(0 0 0 / 5%);\n  margin: 1.5625em 0;\n  color: rgba(0,0,0,0.87);\n  font-size: .85rem;\n">\n  <div style="\n    background-color:rgba(0,184,212,.1);\n    display: flex;\n    align-items: center;\n    padding: .4rem .6rem 0.4rem;\n  ">\n    <span style="fill: rgb(0,184,212); height: 1em; width: 1em; display: inline-flex; margin-right: 0.4rem;">\n      <svg viewBox="0 0 1792 1792" xmlns="http://www.w3.org/2000/svg">\n        <path d="M1152 1376v-160q0-14-9-23t-23-9h-96v-512q0-14-9-23t-23-9h-320q-14 0-23 9t-9 23v160q0 14 9 23t23 9h96v320h-96q-14 0-23 9t-9 23v160q0 14 9 23t23 9h448q14 0 23-9t9-23zm-128-896v-160q0-14-9-23t-23-9h-192q-14 0-23 9t-9 23v160q0 14 9 23t23 9h192q14 0 23-9t9-23zm640 416q0 209-103 385.5t-279.5 279.5-385.5 103-385.5-103-279.5-279.5-103-385.5 103-385.5 279.5-279.5 385.5-103 385.5 103 279.5 279.5 103 385.5z"/>\n</svg>\n    </span>\n    <span style="font-weight: 700;">\n      Trending score\n    </span>\n  </div>\n  <div style="padding: 1.6rem 0.6rem;">\n    \nTo calculate trending score for an event, each bet placed on this event\ncontributes a value determined by how recently the bet was placed.\n\nRecent bets add more value than older ones.\n\nThe normalized value of its score is also assigned to each event as its confidence.\n\n  </div>\n</div>\n\n\n<div style="\n  border-left: 5px solid rgb(124,77,255);\n  border-radius: .1rem;\n  box-shadow: 0 0.2rem 0.5rem rgb(0 0 0 / 5%), 0 0.025rem 0.05rem rgb(0 0 0 / 5%);\n  margin: 1.5625em 0;\n  color: rgba(0,0,0,0.87);\n  font-size: .85rem;\n">\n  <div style="\n    background-color:rgba(124,77,255,.1);\n    display: flex;\n    align-items: center;\n    padding: .4rem .6rem 0.4rem;\n  ">\n    <span style="fill: rgb(124,77,255); height: 1em; width: 1em; display: inline-flex; margin-right: 0.4rem;">\n      <svg viewBox="0 0 1792 1792" xmlns="http://www.w3.org/2000/svg">\n        <path d="M381 1620q0 80-54.5 126t-135.5 46q-106 0-172-66l57-88q49 45 106 45 29 0 50.5-14.5t21.5-42.5q0-64-105-56l-26-56q8-10 32.5-43.5t42.5-54 37-38.5v-1q-16 0-48.5 1t-48.5 1v53h-106v-152h333v88l-95 115q51 12 81 49t30 88zm2-627v159h-362q-6-36-6-54 0-51 23.5-93t56.5-68 66-47.5 56.5-43.5 23.5-45q0-25-14.5-38.5t-39.5-13.5q-46 0-81 58l-85-59q24-51 71.5-79.5t105.5-28.5q73 0 123 41.5t50 112.5q0 50-34 91.5t-75 64.5-75.5 50.5-35.5 52.5h127v-60h105zm1409 319v192q0 13-9.5 22.5t-22.5 9.5h-1216q-13 0-22.5-9.5t-9.5-22.5v-192q0-14 9-23t23-9h1216q13 0 22.5 9.5t9.5 22.5zm-1408-899v99h-335v-99h107q0-41 .5-121.5t.5-121.5v-12h-2q-8 17-50 54l-71-76 136-127h106v404h108zm1408 387v192q0 13-9.5 22.5t-22.5 9.5h-1216q-13 0-22.5-9.5t-9.5-22.5v-192q0-14 9-23t23-9h1216q13 0 22.5 9.5t9.5 22.5zm0-512v192q0 13-9.5 22.5t-22.5 9.5h-1216q-13 0-22.5-9.5t-9.5-22.5v-192q0-13 9.5-22.5t22.5-9.5h1216q13 0 22.5 9.5t9.5 22.5z"/>\n</svg>\n    </span>\n    <span style="font-weight: 700;">\n      Filtering example\n    </span>\n  </div>\n  <div style="padding: 1.6rem 0.6rem;">\n    \nIn this example we get all trending events where `league` is `UEFA Champions League`.\n\n```bash\n$ curl --request GET \\\n  --url \'https://api.vaix.ai/api/sports/events/trending?filters=league%3Aeq%3AUEFA%20Champions%20League\'\n```\n\n  </div>\n</div>\n\n'
        );

        expect(result).toMatchSnapshot();
    });
});
