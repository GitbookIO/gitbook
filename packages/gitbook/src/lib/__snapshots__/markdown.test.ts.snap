// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parseMarkdown should parse a simple table 1`] = `
"<h2>Table</h2>
<table>
<thead>
<tr>
<th>a</th>
<th align="left">b</th>
<th align="right">c</th>
<th align="center">d</th>
</tr>
</thead>
</table>"
`;

exports[`parseMarkdown should parse a complex table 1`] = `
"<p>Returns information for all non-fungible tokens for an account.</p>
<h2>Ordering</h2>
<p>When considering NFTs, their order is governed by a combination of their numerical <strong>token.Id</strong> and <strong>serialnumber</strong> values, with <strong>token.id</strong> being the parent column.
A serialnumbers value governs its order within the given token.id</p>
<p>In that regard, if a user acquired a set of NFTs in the order (2-2, 2-4 1-5, 1-1, 1-3, 3-3, 3-4), the following layouts illustrate the ordering expectations for ownership listing</p>
<ol>
<li><strong>All NFTs in ASC order</strong>: 1-1, 1-3, 1-5, 2-2, 2-4, 3-3, 3-4</li>
<li><strong>All NFTs in DESC order</strong>: 3-4, 3-3, 2-4, 2-2, 1-5, 1-3, 1-1</li>
<li><strong>NFTs above 1-1 in ASC order</strong>: 1-3, 1-5, 2-2, 2-4, 3-3, 3-4</li>
<li><strong>NFTs below 3-3 in ASC order</strong>: 1-1, 1-3, 1-5, 2-2, 2-4</li>
<li><strong>NFTs between 1-3 and 3-3 inclusive in DESC order</strong>: 3-4, 3-3, 2-4, 2-2, 1-5, 1-3</li>
</ol>
<p>Note: The default order for this API is currently DESC</p>
<h2>Filtering</h2>
<p>When filtering there are some restrictions enforced to ensure correctness and scalability.</p>
<p><strong>The table below defines the restrictions and support for the NFT ownership endpoint</strong></p>
<table>
<thead>
<tr>
<th>Query Param</th>
<th>Comparison Operator</th>
<th>Support</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr>
<td>token.id</td>
<td>eq</td>
<td>Y</td>
<td>Single occurrence only.</td>
<td>?token.id=X</td>
</tr>
<tr>
<td></td>
<td>ne</td>
<td>N</td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>lt(e)</td>
<td>Y</td>
<td>Single occurrence only.</td>
<td>?token.id=lte:X</td>
</tr>
<tr>
<td></td>
<td>gt(e)</td>
<td>Y</td>
<td>Single occurrence only.</td>
<td>?token.id=gte:X</td>
</tr>
<tr>
<td>serialnumber</td>
<td>eq</td>
<td>Y</td>
<td>Single occurrence only. Requires the presence of a <strong>token.id</strong> query</td>
<td>?serialnumber=Y</td>
</tr>
<tr>
<td></td>
<td>ne</td>
<td>N</td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>lt(e)</td>
<td>Y</td>
<td>Single occurrence only. Requires the presence of an <strong>lte</strong> or <strong>eq</strong> <strong>token.id</strong> query</td>
<td>?token.id=lte:X&amp;serialnumber=lt:Y</td>
</tr>
<tr>
<td></td>
<td>gt(e)</td>
<td>Y</td>
<td>Single occurrence only. Requires the presence of an <strong>gte</strong> or <strong>eq</strong> <strong>token.id</strong> query</td>
<td>?token.id=gte:X&amp;serialnumber=gt:Y</td>
</tr>
<tr>
<td>spender.id</td>
<td>eq</td>
<td>Y</td>
<td></td>
<td>?spender.id=Z</td>
</tr>
<tr>
<td></td>
<td>ne</td>
<td>N</td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>lt(e)</td>
<td>Y</td>
<td></td>
<td>?spender.id=lt:Z</td>
</tr>
<tr>
<td></td>
<td>gt(e)</td>
<td>Y</td>
<td></td>
<td>?spender.id=gt:Z</td>
</tr>
</tbody>
</table>
<p>Note: When searching across a range for individual NFTs a <strong>serialnumber</strong> with an additional <strong>token.id</strong> query filter must be provided.
Both filters must be a single occurrence of <strong>gt(e)</strong> or <strong>lt(e)</strong> which provide a lower and or upper boundary for search.</p>"
`;
